Add support for event schema validation

yagnik opened the pull request:
### What does this PR do?
Add support for event schema validation

### What issues does this PR fix or reference?
closes https://github.com/saltstack/salt/issues/41896

### Previous Behavior
All events were fired and received without any schema validation

### New Behavior
Every event needs to have a schema and validated before getting on the event bus and at read time

### Tests written?
Yes

### Questions/Concerns:
- The request adds jsonschema in base requirement, do we want it to be optional ?
- The loading mechanism for schema uses lazy loader to keep a consistent behaviour across loading but the loader is present in the hot path of event pub/sub, I want to avoid hardcoding events into a list so that the user can extend the schema to add custom schema and use the message bus for their own purpose, do we have a better way to handle this ?

yagnik committed:
Add support for event-schema validation

salt-jenkins commented:
@yagnik, thanks for your PR! By analyzing the history of the files in this pull request, we identified @terminalmage, @jacksontj and @skizunov to be potential reviewers.

yagnik commented on salt/loader.py:
```diff
@@ -444,6 +444,15 @@ def auth(opts, whitelist=None):
     )
 
 
+def schemas(opts, whitelist=None):
```
I have mixed feelings on this, I like that it is consistent with all other loaders but given that this is now in the hot path of every event being emitted, it may cause performance issues.

thatch45 replied:
Using a loader here is the right approach, this does not cause me concern

yagnik commented on salt/utils/schema.py:
```diff
@@ -687,6 +688,15 @@ def as_requirements_item(cls):
+    tag = None
+
+    @classmethod
+    def match(cls, tag):
```
by default we use regex with tag matches but can be overwritten for tags where the minion id is present in the tag. 
Also currently this regex match is way to open, I'll update it to be stricter if you all agree.

cachedout commented:
Hi @yagnik 

I really like the direction this is going but I think that we need to think carefully about what external library to use. We tend to be very reluctant to add additional dependencies to Salt and this is no exception.

We have recently been exploring the use of Toasted Marshmellow for this purpose. Have a look and let me know what you think: https://github.com/lyft/toasted-marshmallow (Thanks to @ryan-lane for alerting us to this project from our friends at Lyft.)

yagnik commented:
100% in favour of not adding any extra dependency if possible to salt. 
Currently schema was built using python version of json schema but I didn't see it being actively used anywhere that would have a serious impact. So changing the schema library should be possible. 

Having said the above, from a cursory look we will still have to add another dependency whether it be toastedmarshmallow or jsonschema. Currently I'm in favour of json schema for the following reasons:
- dependency needs to be added for both libraries
- json schema is more accepted spec than marshmallow
- the dev activity on marshmallow seems slower and the two layer projects gets me concerned about patches and updates
- unsure about the speed of validation yet compared to marshmallow

Things I like about marshmallow:
- Converting a schema into an object is really nice but we should be able to do the same with a decorator maybe ?

What do you think about the above @cachedout  ?

cachedout commented:
@yagnik Have you had a look at what's currently in `salt.utils.schema` and `salt.config.schemas`? I wonder if we could leverage that work?

yagnik commented:
@cachedout yep I did. The work above is based off from that code. 
We don't actually validate any schema right now from what I could understand, it was made for future proofing but the validation requires jsonschema: https://github.com/saltstack/salt/blob/develop/salt/utils/schema.py#L120

yagnik commented:
Any new thoughts on this ? Can I help clarify anything ?

cachedout commented:
@yagnik I was just looking at this yesterday with @thatch45. I think he was going to comment here but got distracted. I will remind him again today. Thanks for the reminder.

thatch45 commented on requirements/base.txt:
```diff
@@ -6,3 +6,4 @@ requests>=1.0.0
 tornado>=4.2.1
 # Required by Tornado to handle threads stuff.
 futures>=2.0
+jsonschema==2.6.0
```
Since this is disabled by default I don't think jsonschema needs to be a required base dep

cachedout replied:
Totally agree. We can come back to this question in future releases if this takes off but for the time being, I don't feel like we can accept this as a hard dependency for Salt.

thatch45 commented on salt/utils/event.py:
```diff
@@ -873,6 +882,36 @@ def __del__(self):
+
+    def _find_schema_by_tag(self, tag):
+        all_schemas = []
+        for schema_mods in salt.loader.schemas(self.opts):
```
We should not invoke the loader like this, this is causing the loader to re-scan with every event and will be a nightmare.
We should make an object level instance of the schema loader object and then use it.

thatch45 reviewed (changes requested):
In a nutshell I am on board with the approach here, we just need to fix a few things with how the loader is being used

terminalmage reviewed (changes requested):
I concur with @thatch45's points.

yagnik commented on salt/utils/event.py:
```diff
@@ -873,6 +885,36 @@ def __del__(self):
+
+    def _find_schema_by_tag(self, tag):
+        if len(SaltEvent.SCHEMAS) == 0:
+            for schema_mods in salt.loader.schemas(self.opts):
```
I decided to make class level schemas list and memoize it because I wasn't able to find a nice way to access opts before class is loaded. Any better way to do this ?

yagnik commented:
Updated the code to memoize schema loading and removed json schema from base. Please do let me know if you think I can do the memoization of schemas better.

rallytime commented:
Re-review from @thatch45 and @terminalmage requested.

thatch45 commented on salt/utils/event.py:
```diff
@@ -873,6 +886,36 @@ def __del__(self):
+
+    def _find_schema_by_tag(self, tag):
+        if len(SaltEvent.SCHEMAS) == 0:
+            for schema_mods in salt.loader.schemas(self.opts):
```
Why are we calling the loader here? We should call call the loader when we init the class, I find this module level object to be very odd.

yagnik replied:
You are correct, will fix this.

thatch45 reviewed (approved):
I think I might be ok with this approach, it is just different from what I am used to using in Salt

yagnik commented:
@thatch45 would you have an example of how you would use in salt ? I don't mind updating the code to follow the salt guidelines.

thatch45 commented on salt/utils/event.py:
```diff
@@ -210,6 +217,9 @@ class SaltEvent(object):
+    SCHEMAS = []
+
     def __init__(
             self, node, sock_dir=None,
```
```
if self.opts['schemas_enabled']:
    self.schemas = salt.loader.schemas(self.opts)
else:
    self.schemas = {}
```
so typically at class instantiation time instead of dynamically updating a class level list inside the `__init__`. That way you just have an assurance that the schemas are created and don't need to double check during runtime.

yagnik replied:
I definitely like this but the reason for doing class level was because I didn't want to load schemas every time event object is created. This way a salt process doesn't reload schemas every time instance is created which during the life time of the process would be many times. wdyt ?

thatch45 commented:
Sure thing, I left a comment inline @yagnik :)

thatch45 commented:
I think that the per instance approach is still better because we have a local instance in memory to track and use, also we usually only have a single event connection per process

yagnik commented:
Updated code based on comments, should be good to go now :)

yagnik commented:
@terminalmage can I get your blessings too ?

isbm commented:
I have already done something that direction, experimenting with structures. Problem I am trying to solve is that in Salt everyone can return anything at any time. Which is cool for hacking, but doesn't really scale in a long run. APIs aren't that stable, so we are thinking how to be nice to developers and keep it simple, but also do not ruin APIs and be backward-compatible.

So here my differences are the following:

1. There is a place where structures/schemas are defined per a module/method and they are just a hash/JSON. This stuff is automatically resolved by a lazyloader and processed with the decorator later. For example:
```python
class Package(object):
    '''
    Refers to the pkg.* module.
    '''
    class _Methods(object):
        __metaclass__ = ABCMeta

        @abstractmethod
        def mod_repo(self, data):
            '''
            Structure returned for mod_repo method call.
            '''
            schema = {
                'alias' : {
                    'baseurl': str,
                    'enabled': bool,
                    'refresh': bool,
                    'gpgcheck': bool,
                    'type': str,
                    'filename': str,
                }
            }
            return schema
```

2. I use decorators. They allow to avoid touching the code and plus they allow flowing "old" data between modules and states (this way we do not rewrite anything there). Basically, to use the above, I have the following code (again, in my branch still) e.g. for `salt/modules/yumpkg.py`:
```python
from salt.utils import scm
...
@scm.pkg(__file__)
def mod_repo(repo, basedir=None, **kwargs):
....
```

Then for `def mod_repo` you just set a decorator, which will verify the output according to the structure. However, this decorator is disabled once state calls a module, so it is using old data structure. The result is that the output of the module functions are expected and can be questioned for something like "What structure I should expect, once I call it?".

This my branch is actually still works, but it was kind of suspended for a long time and I will need to resume work on it (as per discussion with @cachedout and @thatch45 last week).

isbm commented:
@yagnik so what I am saying is: maybe we could join the forces here and think twice before we go full-blown classes? I am using classes as a container and structure to reflect the modules and states, whether the schema is just a dict (and it is anyway in Salt internals). I would open a branch on our https://github.com/opensuse/salt and you could join and we could work together then, and so then we could push all that here afterwards.

What you think?

yagnik commented:
@isbm I was initially thinking about just caring about schemas that go on the bus but this makes a lot of sense in terms of what the code expects. 
Let me know how I can help and we can discuss the class approach further too. 
Closing this request for now.

isbm commented:
@yagnik wait please few days, I will open a branch for you (I mean, us), and we will be able to put all that together. Let me ping you here soon.
